### PR TITLE
OPSEXP-1565 Use pipenv for dependency management of playbook

### PR DESCRIPTION
**Pull request by piraartur**
OPSEXP-1565

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -1,24 +1,24 @@
 Vagrant.configure("2") do |config|
+  config.vm.provision "file", source: "~/Pipfile", destination: "$HOME/vagrant"
```
Comment (gionn) [inline Vagrantfile:2]: `~/Pipfile`? this is pointing to a Pipfile in your mac home

File: Vagrantfile [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
-    ansible.install_mode = "pip_args_only"
-    ansible.pip_args = "-r /vagrant/requirements.txt"
-    ansible.pip_install_cmd = "sudo apt-get install -y python3-pip python-is-python3 haveged && sudo ln -s -f /usr/bin/pip3 /usr/bin/pip"
     ansible.galaxy_role_file = "requirements.yml"
     ansible.extra_vars = {
       nexus_user: ENV['NEXUS_USERNAME'],
       nexus_password: ENV['NEXUS_PASSWORD'],
       autogen_unsecure_secrets: true
     }
   end
```
Comment (gionn) [inline Vagrantfile:23]: as already said, I don't think we can use this plugin when using pipenv, please remove this section

File: install_pipenv.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+apt-get update
+apt-get install python3.10 -y
```
Comment (gionn) [inline install_pipenv.sh:3]: python should be already present on the base box

File: install_pipenv.sh [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+apt-get install python3-pip -y
+pip3 install -U urllib3 requests
+pip3 install --user pipenv
+echo -e 'export PATH="$PATH:/Library/Frameworks/Python.framework/Versions/3.9/bin"' >> $HOME/.bash_profile
```
Comment (gionn) [inline install_pipenv.sh:7]: this is a macos path and doesn't make sense in the vagrant box that is ubuntu

**Review by piraartur — COMMENTED** [review]
I have forgot to clean block of code from ACS playbook, thanks to @kklassa who spotted it when he saw my PR ready to review.

File: .github/workflows/enteprise.yml [hunk 5]
```diff
@@ -150,4 +154,4 @@ jobs:
           timeout_minutes: 20
           max_attempts: 3
           retry_wait_seconds: 60
-          command: ./tests/molecule_it/script.sh destroy
+          command: ./tests/molecule_it/script.sh destroy
```
Comment (gionn) [inline .github/workflows/enteprise.yml:157]:
```suggestion
          command: ./tests/molecule_it/script.sh destroy

```

File: Pipfile [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+molecule = "==4.0.1"
+molecule-docker = "==2.0.0"
+molecule-ec2 = "==0.4"
+boto3 = "==1.24.3"
+pyhamcrest = "==2.0.3"
+pylint = "==2.14.5"
+pytest-testinfra = "==6.8.0"
+requests = "==2.28.1"
+yamllint = "==1.27.1"
+dnspython = "==2.2.1"
```
Comment (gionn) [inline Pipfile:19]: move those development packages (they are not required to run ansible itself) to the dev-packages section below

then you need to add `--dev` to the pipenv install command

File: .github/workflows/community.yml [hunk 7]
```diff
@@ -39,7 +39,8 @@ jobs:
           python-version-file: '.python-version'
           cache: "pip"
 
-      - run: pip install -r requirements.txt
+      - run: pip install --user pipenv
+      - run: pipenv install
```
Comment (gionn) [inline .github/workflows/community.yml:43]: I think it make sense for the CI to use `--deploy` option, see https://pipenv-fork.readthedocs.io/en/latest/basics.html#pipenv-install

File: .github/workflows/enteprise.yml [hunk 8]
```diff
@@ -32,7 +32,8 @@ jobs:
           python-version-file: '.python-version'
           cache: "pip"
 
-      - run: pip install -r requirements.txt
+      - run: pip install --user pipenv
+      - run: pipenv install
```
Comment (gionn) [inline .github/workflows/enteprise.yml:36]: if you feel adventurous, you can try making a new composite github action that includes those two run commands and the previous setup python, like we already did for ./.github/actions/galaxy below

in this way you don't duplicate commands in all the workflows

File: .github/workflows/enteprise.yml [hunk 9]
```diff
@@ -133,7 +135,9 @@ jobs:
           AWS_ACCESS_KEY_ID: ${{ secrets.aws_access_key_id }}
           AWS_SECRET_ACCESS_KEY: ${{ secrets.aws_secret_access_key }}
           CLONE_GITHUB_TOKEN: ${{ secrets.BOT_GITHUB_TOKEN }}
-        run: ./tests/molecule_it/script.sh verify
+        run: |
+         pipenv run molecule
+         ./tests/molecule_it/script.sh verify
```
Comment (gionn) [inline .github/workflows/enteprise.yml:140]: is this a leftover? revert it

File: Vagrantfile [hunk 10]
```diff
@@ -11,14 +12,11 @@ Vagrant.configure("2") do |config|
...
     ansible.become = true
-    ansible.install_mode = "pip_args_only"
-    ansible.pip_args = "-r /vagrant/requirements.txt"
-    ansible.pip_install_cmd = "sudo apt-get install -y python3-pip python-is-python3 haveged && sudo ln -s -f /usr/bin/pip3 /usr/bin/pip"
     ansible.galaxy_role_file = "requirements.yml"
     ansible.extra_vars = {
       nexus_user: ENV['NEXUS_USERNAME'],
       nexus_password: ENV['NEXUS_PASSWORD'],
       autogen_unsecure_secrets: true
     }
```
Comment (gionn) [inline Vagrantfile:20]: ansible_local provision should be unnecessary now - remove it

File: docs/deployment-guide.md [hunk 11]
```diff
@@ -180,13 +180,14 @@ Not all distributions of Linux may have that specific version of Ansible. Below
...
-3. Install Ansible and required dependencies in python virtualenv:
+3. Install Ansible and required dependencies while also creating virtual environment:
 
     ```bash
     unzip alfresco-ansible-deployment-<version>.zip
     cd alfresco-ansible-deployment
     mkvirtualenv alfresco-ansible
-    pip install -r requirements.txt
+    pip install --user pipenv
+    pipenv install Pipfile
```
Comment (gionn) [inline docs/deployment-guide.md:190]:
```suggestion
    pipenv install
```

**Review by gionn — CHANGES_REQUESTED** [review]

File: Vagrantfile (hunk 10 above)
Comment (piraartur) [reply Vagrantfile:20]: > ansible_local provision should be unnecessary now - remove it

What about ansible.extra_vars? nexus credentials?

File: .github/workflows/enteprise.yml (hunk 9 above)
Comment (piraartur) [reply .github/workflows/enteprise.yml]: I do believe it is necessary, because after deleting pipenv run molecule, it stopped working

Comment (gionn) [reply .github/workflows/enteprise.yml:140]: trust the build output, it's really just printing the command help 😆 
https://github.com/Alfresco/alfresco-ansible-deployment/runs/7746557184?check_suite_focus=true#step:11:27

File: playbooks/acs.yml [hunk 12]
```diff
@@ -1,4 +1,20 @@
...
+      shell:  | 
+        sudo su
+        add-apt-repository ppa:deadsnakes/ppa
+        apt-get update
+        apt-get install python3.9 -y
+        apt-get install python3-pip -y
+        pip3 install -U urllib3 requests
+        python3.9 -m pip install pipenv
+        pipenv install
+
```
Comment (gionn) [inline playbooks/acs.yml:17]:
```suggestion
```

not needed

**Review by gionn — CHANGES_REQUESTED** [review]

File: docs/deployment-guide.md [hunk 13]
```diff
@@ -180,13 +180,14 @@ Not all distributions of Linux may have that specific version of Ansible. Below
...
-3. Install Ansible and required dependencies in python virtualenv:
+3. Install Ansible and required dependencies while also creating virtual environment:
 
     ```bash
     unzip alfresco-ansible-deployment-<version>.zip
     cd alfresco-ansible-deployment
     mkvirtualenv alfresco-ansible
-    pip install -r requirements.txt
+    pip install --user pipenv
+    pipenv install
```
Comment (gionn) [inline docs/deployment-guide.md:190]:
```suggestion
    pipenv install --deploy
```

File: .github/workflows/enteprise.yml [hunk 14]
```diff
@@ -133,7 +135,7 @@ jobs:
           AWS_ACCESS_KEY_ID: ${{ secrets.aws_access_key_id }}
           AWS_SECRET_ACCESS_KEY: ${{ secrets.aws_secret_access_key }}
           CLONE_GITHUB_TOKEN: ${{ secrets.BOT_GITHUB_TOKEN }}
-        run: ./tests/molecule_it/script.sh verify
+        run: pipenv run ./tests/molecule_it/script.sh verify
```
[the remark this answers is not in the document]
Comment (piraartur) [reply .github/workflows/enteprise.yml]: It is, in commits prior to this the workflow was failing/printing molecule default usage guide or just skipped script

**Review by gionn — APPROVED** [review]

**Pull request merged**